internal/impl/babysitter.go

```diff
@@ -142,21 +164,80 @@ func RunBabysitter(ctx context.Context) error {
 
 // ActivateComponent implements the envelope.EnvelopeHandler interface.
 func (b *babysitter) ActivateComponent(_ context.Context, request *protos.ActivateComponentRequest) (*protos.ActivateComponentReply, error) {
-	// Got a request from the weavelet to activate a component. However, the
-	// component should be already activated.
-	rs := replicaSet(request.Component, b.cfg.Deployment)
+	go func() {
+		if err := b.watchPods(b.ctx, request.Component); err != nil {
+			// TODO(mwhittaker): Log this error.
+			fmt.Fprintf(os.Stderr, "watchPods(%q): %v", request.Component, err)
+		}
+	}()
+	return &protos.ActivateComponentReply{}, nil
+}
 
-	// Tell the weavelet the address of the service that runs the component.
+// watchPods watches the pods hosting the provided component, updating the
+// routing info whenever the set of pods changes.
+func (b *babysitter) watchPods(ctx context.Context, component string) error {
+	b.mu.Lock()
+	if _, ok := b.watching[component]; ok {
+		// The pods for this component are already being watched.
+		b.mu.Unlock()
+		return nil
+	}
+	b.watching[component] = struct{}{}
+	b.mu.Unlock()
+
+	// Watch the pods running the requested component.
+	rs := replicaSet(component, b.cfg.Deployment)
 	name := name{b.cfg.Deployment.App.Name, rs, b.cfg.Deployment.Id[:8]}.DNSLabel()
-	addr := fmt.Sprintf("tcp://%s:%d", name, servicePort)
-	routingInfo := &protos.RoutingInfo{
-		Component: request.Component,
-		Replicas:  []string{addr},
+	opts := metav1.ListOptions{LabelSelector: fmt.Sprintf("depName=%s", name)}
+	watcher, err := b.clientset.CoreV1().Pods("default").Watch(ctx, opts)
+	if err != nil {
+		return fmt.Errorf("watch pods for component %s: %w", component, err)
 	}
-	if err := b.envelope.UpdateRoutingInfo(routingInfo); err != nil {
-		return nil, err
+
+	addrs := map[string]string{}
```

Comment:
can you add a comment for the big for loop?

Reply:
Done.